src/main/java/uk/gov/ons/census/action/schedule/ActionInstructionBuilder.java

```diff
@@ -0,0 +1,148 @@
+    ActionEvent actionEvent = new ActionEvent();
+    actionEvent
+        .getEvents()
+        .add("CASE_CREATED : null : SYSTEM : Case created when Initial creation of case");
+
+    ActionAddress actionAddress = new ActionAddress();
+    actionAddress.setLine1(caze.getAddressLine1());
+    actionAddress.setLine2(caze.getAddressLine2());
+    actionAddress.setLine3(caze.getAddressLine3());
+    actionAddress.setTownName(caze.getTownName());
+    actionAddress.setPostcode(caze.getPostcode());
+    actionAddress.setOrganisationName(caze.getOrganisationName());
+    ActionRequest actionRequest = new ActionRequest();
+    actionRequest.setActionId(UUID.randomUUID().toString());
+    actionRequest.setResponseRequired(false);
+    actionRequest.setActionPlan(actionRule.getActionPlan().getId().toString());
+
+    actionRequest.setActionType(actionRule.getActionType().toString());
+
+    actionRequest.setAddress(actionAddress);
+    actionRequest.setLegalBasis("Statistics of Trade Act 1947");
+    actionRequest.setCaseGroupStatus("NOTSTARTED");
+    actionRequest.setCaseId(caze.getCaseId().toString());
+
+    actionRequest.setPriority(Priority.MEDIUM);
+    actionRequest.setCaseRef(Long.toString(caze.getCaseRef()));
+    actionRequest.setIac(uacQidTuple.getUacQidLink().getUac());
+    actionRequest.setQid(uacQidTuple.getUacQidLink().getQid());
+
+    if (uacQidTuple.getUacQidLinkWales().isPresent()) {
+      actionRequest.setIacWales(uacQidTuple.getUacQidLinkWales().get().getUac());
+      actionRequest.setQidWales(uacQidTuple.getUacQidLinkWales().get().getQid());
+    }
+
+    actionRequest.setEvents(actionEvent);
+    actionRequest.setExerciseRef("201904");
+    actionRequest.setUserDescription("Census-FNSM580JQE3M4");
+    actionRequest.setSurveyName("Census-FNSM580JQE3M4");
+    actionRequest.setSurveyRef("Census-FNSM580JQE3M4");
+    actionRequest.setReturnByDate("27/04");
+    actionRequest.setSampleUnitRef("DDR190314000000516472");
+    ActionInstruction actionInstruction = new ActionInstruction();
+    actionInstruction.setActionRequest(actionRequest);
+
+    return actionInstruction;
+  }
+
+  private boolean isQuestionnaireWelsh(String treatmentCode) {
+    return (treatmentCode.startsWith(HOUSEHLD_INITIAL_CONTACT_QUESTIONNIARE_TREATMENT_CODE_PREFIX)
+        && treatmentCode.endsWith(WALES_TREATMENT_CODE_SUFFIX));
+  }
+
+  private UacQidTuple getUacQidLinks(Case caze) {
+    List<UacQidLink> uacQidLinks = uacQidLinkRepository.findByCaseId(caze.getCaseId().toString());
+    UacQidTuple uacQidTuple = new UacQidTuple();
+
+    if (uacQidLinks == null || uacQidLinks.isEmpty()) {
+      throw new RuntimeException(); // TODO: How can we process this case without a UAC?
+    } else if (uacQidLinks.size() > THE_NUMBER_ONE_IS_NOW_A_CONSTANT_HURRAH_THIS_IS_MUCH_BETTER) {
+      if (isQuestionnaireWelsh(caze.getTreatmentCode())
+          && uacQidLinks.size()
+              == NUM_OF_UAC_IAC_PAIRS_NEEDED_BY_A_WALES_INITIAL_CONTACT_QUESTIONNAIRE) {
```

Comment:
Perhaps this constant could be a bit more verbose,  for example NUM is an abbreviations  and UAC / IAC are acronyms that could be fully expanded here to aid comprehension

Reply:
good idea